robot/impl/resource_manager.go

```diff
@@ -529,6 +530,152 @@ func (manager *resourceManager) completeConfig(
+					return
+				}
+
+				var verb string
+				conf := gNode.Config()
+				if gNode.IsUninitialized() {
+					verb = "configuring"
+					gNode.InitializeLogger(
+						manager.logger, resName.String(), conf.LogConfiguration.Level,
+					)
+				} else {
+					verb = "reconfiguring"
+				}
+				manager.logger.CInfow(ctx, fmt.Sprintf("Now %s resource", verb), "resource", resName)
+
+				// this is done in config validation but partial start rules require us to check again
+				if _, err := conf.Validate("", resName.API.Type.Name); err != nil {
+					gNode.LogAndSetLastError(
+						fmt.Errorf("resource config validation error: %w", err),
+						"resource", conf.ResourceName(),
+						"model", conf.Model)
+					return
+				}
+				if manager.moduleManager.Provides(conf) {
+					if _, err := manager.moduleManager.ValidateConfig(ctxWithTimeout, conf); err != nil {
+						gNode.LogAndSetLastError(
+							fmt.Errorf("modular resource config validation error: %w", err),
+							"resource", conf.ResourceName(),
+							"model", conf.Model)
+						return
+					}
+				}
+
+				switch {
+				case resName.API.IsComponent(), resName.API.IsService():
+					processResource := func() {
+						defer levelWG.Done()
+
+						newRes, newlyBuilt, err := manager.processResource(ctxWithTimeout, conf, gNode, lr)
+						if newlyBuilt || err != nil {
+							if err := manager.markChildrenForUpdate(resName); err != nil {
+								manager.logger.CErrorw(ctx,
+									"failed to mark children of resource for update",
+									"resource", resName,
+									"reason", err)
+							}
+						}
+
+						if err != nil {
+							gNode.LogAndSetLastError(
+								fmt.Errorf("resource build error: %w", err),
+								"resource", conf.ResourceName(),
+								"model", conf.Model)
+							return
+						}
+
+						// if the ctxWithTimeout fails with DeadlineExceeded, then that means that
+						// resource generation is running async, and we don't currently have good
+						// validation around how this might affect the resource graph. So, we avoid
+						// updating the graph to be safe.
+						if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
+							manager.logger.CErrorw(
+								ctx, "error building resource", "resource", conf.ResourceName(), "model", conf.Model, "error", ctxWithTimeout.Err())
+						} else {
+							gNode.SwapResource(newRes, conf.Model)
+						}
+					}
+
+					if !forceSync {
+						// TODO(RSDK-6925): support concurrent processing of resources of
+						// APIs with a maximum instance limit. Currently this limit is
+						// validated later in the resource creation flow and assumes that
+						// each resource is created synchronously to have an accurate
+						// creation count.
+						if c, ok := resource.LookupGenericAPIRegistration(resName.API); ok && c.MaxInstance != 0 {
+							forceSync = true
+						}
+					}
+
+					levelWG.Add(1)
+					if forceSync {
+						processResource()
+					} else {
+						go processResource()
```

Comment:
I think this needs to be wrapped in a `lr.reconfigureWorkers.Add` + `Done`.

Reply:
ah you're right - this process can leak beyond the lifetime of the `goutils.PanicCapturingGo`-routine now.

i wonder if we can also just `reconfigureWorkers` instead of the local `levelWG` to control this flow - it's currently just being used for tests.

Reply:
done https://github.com/viamrobotics/rdk/pull/3707/commits/072026617d3518f2c027921c2f8b798aadbe3521